src/mqtt/device_mqtt_impl.cpp

```diff
@@ -0,0 +1,100 @@
+// (C) Copyright 2025, SECO Mind Srl
+//
+// SPDX-License-Identifier: Apache-2.0
+
+// macro to mark unimplemented code. Throws an exception with file/line info.
+#define TODO(msg)                                                                \
+  throw std::logic_error("TODO: " + std::string(msg) + " at " + __FILE__ + ":" + \
+                         std::to_string(__LINE__));
+
+#include "mqtt/device_mqtt_impl.hpp"
+
+#include <chrono>
+#include <filesystem>
+#include <list>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+
+#include "astarte_device_sdk/data.hpp"
+#include "astarte_device_sdk/device_mqtt.hpp"
+#include "astarte_device_sdk/mqtt/config.hpp"
+#include "astarte_device_sdk/mqtt/pairing.hpp"
+#include "astarte_device_sdk/msg.hpp"
+#include "astarte_device_sdk/object.hpp"
+#include "astarte_device_sdk/ownership.hpp"
+#include "astarte_device_sdk/property.hpp"
+#include "astarte_device_sdk/stored_property.hpp"
+
+namespace AstarteDeviceSdk {
+
+AstarteDeviceMQTT::AstarteDeviceMQTTImpl::AstarteDeviceMQTTImpl(const MqttConfig cfg)
+    : cfg_(std::move(cfg)) {}
+
+AstarteDeviceMQTT::AstarteDeviceMQTTImpl::~AstarteDeviceMQTTImpl() = default;
+
+void AstarteDeviceMQTT::AstarteDeviceMQTTImpl::add_interface_from_file(
+    const std::filesystem::path& json_file) {
+  TODO("not yet implemented");
+}
+void AstarteDeviceMQTT::AstarteDeviceMQTTImpl::add_interface_from_str(std::string_view json) {
+  TODO("not yet implemented");
+}
+void AstarteDeviceMQTT::AstarteDeviceMQTTImpl::remove_interface(const std::string& interface_name) {
+  TODO("not yet implemented");
+}
+
+void AstarteDeviceMQTT::AstarteDeviceMQTTImpl::connect() {
+  // FIXME: at the moment this method just serves to retrieve the credential secret from Astarte,
+  // so that the device appears to be registered to astarte. It does not establishing an effective
+  // connection to Astarte.
+  auto pairing_api = PairingApi(cfg_.realm(), cfg_.device_id(), cfg_.pairing_url());
+  pairing_api.register_device(cfg_.pairing_token().value());
+  connected_ = true;
```

Comment:
This is an issue in my opinion. We have two possible approaches for this library:
- Keep the device registration as an external functionality (the separate PairingAPI class) that the users can call before attempting the connection. In this way the user has control over where the credential secret should be stored in persistency.
- Move the device registration within the device class. In that case we should provide the user some optional callback or context that allows them to decide where the credential secret should be stored in persistency.

All this with the objective of offloading the "safety" aspect of storing a credential secret. As it's very application dependent.